---
title: "R Review"
output: html_notebook
date: 11/15/24
---

Part 1: Review Questions

1. TCGA is the Cancer Genome Atlas and it is important because it contains comprehensive, publicly accessible genomic data with over 20,000 samples across 33 cancer types. The dataset is a valuable resource for identifying genetic mutations and pathways for research and treatment methods.

2. Some strengths of TCGA is that it contains a large amount of data across different omics: genomic, transcriptomic, and epigenomic data. The weakness of TCGA is that it includes limited representation of rare cancers and may contain potential demographic biases which in turn does not encapsulate the genetic diversity of a whole population. 

3. Git status, git add <filename>, git commit -m “done”, git push

4. install.packages("package") → library(package)

5. BiocManager::install() →  library(BiocManager)

6. Boolean indexing is a very effective R technique for data cleaning, subsetting or selection, and conditional analysis. Boolean indexing involves applying a vector of booleans (true or false) to a column or row in a dataframe based on a particular condition. 

7. Sample Data Frame
```{r}
students <- data.frame(Name = c("Alice", "Bob", "Charlie", "Diana", "Evan"),
                       Score = c(55, 92, 76, 61, 45), 
                       Age = c(20, 21, 22, 19, 20)) 
students

# ifelse() statement
students$pass_or_fail <- ifelse(students$Score <= 60, "Fail", "Pass")

# Boolean Indexing
passed_student_index <- students[students$pass_or_fail == "Pass",]

```
Part 2: SKCM Analysis
```{r setup}
knitr::opts_knit$set(root.dir = normalizePath("/home1/aditis54/490_cluster/analysis_data"))
```

```{r}
library(maftools) 
library(TCGAbiolinks)
library(BiocManager)
library(survival)
library(survminer)
library(SummarizedExperiment)

```

```{r}
clin_query <- GDCquery(project = "TCGA-SKCM", 
                       data.category = "Clinical", 
                       data.type = "Clinical Supplement", 
                       data.format = 'BCR Biotab') 

GDCdownload(clin_query)
clinical.BCRtab.all <- GDCprepare(clin_query)
```

```{r}
clinical <- read.csv("skcm_clinical_data.csv")
rad <- clinical.BCRtab.all$clinical_radiation_skcm[-c(1,2),]
drug <- clinical.BCRtab.all$clinical_drug_skcm[-c(1,2),]

```

1. Difference in survival between metastatic and non-metastatic patients (KM plot)
```{r}
rna_query <- GDCquery(project ="TCGA-SKCM",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")

GDCdownload(rna_query)
rna_se <- GDCprepare(rna_query)
```

```{r}
rna_clinical = read.csv("skcm_rna_clinical.csv")
rna_clinical$Tumor_Sample_Barcode <- rna_clinical$patient
```

```{r}
days_to_death_NA_mask <- !is.na(rna_clinical$days_to_death)
survival_rna_clinical <- rna_clinical[days_to_death_NA_mask,]
survival_rna_clinical$vital_status_num <- as.numeric(survival_rna_clinical$vital_status == "Dead")
```

```{r}
survival_object <- Surv(time = survival_rna_clinical$days_to_death, 
                        event = survival_rna_clinical$vital_status_num)

fit_object <- survfit(survival_object ~ survival_rna_clinical$definition, 
                      data = survival_rna_clinical)
```

```{r}
survplot <- ggsurvplot(fit_object, pval = TRUE,
                       ggtheme = theme(plot.margin 
                                       = unit(c(1,1,1,1), "cm")), 
                       legend = "right")

KM_plot <- survplot$plot + theme_bw() + 
  theme(axis.title = element_text(size=20),
        axis.text = element_text(size=16),
        legend.title = element_text(size=14),
        legend.text = element_text(size=12))

KM_plot
```
2. Differential expression between non-metastatic and metastatic patients controlling for treatment effects, race, gender, and vital status (DESeq2 + Volcano plot)

a. Treatments must include radiation, chemotherapy, immunotherapy, molecular therapy, vaccine
```{r}
if (!require("DESeq2", quietly = TRUE))
BiocManager::install("DESeq2")
BiocManager::install("EnhancedVolcano") 
library(EnhancedVolcano)
```

```{r}
rna_clinical <- rna_clinical[, -33] # removes treatments
rna_genes <- read.csv("skcm_rna_genes.csv")
row.names(rna_genes) <- rna_genes$gene_id # add descriptive row name
```
```{r}
rna_counts <- read.csv("skcm_rna_counts.csv")
row.names(rna_counts) <- rna_genes$gene_id # add descriptive row name
colnames(rna_counts) <- rna_clinical$barcode
```
```{r}
# variable of interest
definition_mask <- ifelse(rna_clinical$definition == "Additional Metastatic" | rna_clinical$definition == "Solid Tissue Normal", F, T)
rna_clinical <- rna_clinical[definition_mask, ]
rna_clinical$definition <- factor(rna_clinical$definition)
```
```{r}
# covariates

# For Chemotherapy

chemo <- drug[drug$pharmaceutical_therapy_type == 'Chemotherapy',]
rna_clinical$chemotherapy <- ifelse(rna_clinical$patient %in% chemo$bcr_patient_barcode, 1, 0)

# immunotherapy
immuno <- drug[drug$pharmaceutical_therapy_type == 'Immunotherapy',]
rna_clinical$immunotherapy <- ifelse(rna_clinical$patient %in% immuno$bcr_patient_barcode, 1, 0)

# radiation
rna_clinical$radiation <- ifelse(rna_clinical$patient %in% rad$bcr_patient_barcode, 1, 0)

# molecular
molec <- drug[drug$pharmaceutical_therapy_type == 'Targeted Molecular therapy',]
rna_clinical$molecular <- ifelse(rna_clinical$patient %in% molec$bcr_patient_barcode, 1, 0)

# vaccine
vaccine <- drug[drug$pharmaceutical_therapy_type == 'Vaccine',]
rna_clinical$vaccine <- ifelse(rna_clinical$patient %in% vaccine$bcr_patient_barcode, 1, 0)

# Factorize all columns
rna_clinical$chemotherapy <- factor(rna_clinical$chemotherapy)
rna_clinical$immunotherapy <- factor(rna_clinical$immunotherapy)
rna_clinical$radiation <- factor(rna_clinical$radiation)
rna_clinical$molecular <- factor(rna_clinical$molecular)
rna_clinical$vaccine <- factor(rna_clinical$vaccine)

# Other covariates: race, gender, and vital status

NA_race_mask <- ifelse(rna_clinical$race == "not reported", F, T)
rna_clinical <- rna_clinical[NA_race_mask, ]
rna_clinical$race <- factor(rna_clinical$race)

rna_clinical$gender <- factor(rna_clinical$gender)

NA_vital_status_mask <- ifelse(rna_clinical$vital_status == "Not Reported", F, T)
rna_clinical <- rna_clinical[NA_vital_status_mask, ]
rna_clinical$vital_status <- factor(rna_clinical$vital_status)

# matching the clinical, counts, and genes dataframes
rna_counts <- rna_counts[, rna_clinical$barcode, drop = FALSE]
# rna_counts variables should match rna_clinical obs

less_10_mask <- rowSums(rna_counts) >= 10
rna_counts <- rna_counts[less_10_mask, ]
rna_genes <- rna_genes[less_10_mask, ]
```
```{r}
# Perform the differential expression analysis

dds <- DESeqDataSetFromMatrix(countData = rna_counts,
                              colData = rna_clinical,
                              design=~chemotherapy + immunotherapy + radiation + molecular + vaccine + race + gender + vital_status + definition)

dds_obj <- DESeq(dds)

```
```
```{r setup}
knitr::opts_knit$set(root.dir = normalizePath("/home1/aditis54/490_cluster/analysis_data"))
```

```{r}
results <- read.csv('/home1/aditis54/490_cluster/analysis_data/SKCM_DESeq2_metastasis_results.csv', row.names=1)
results <- data.frame(results)

# Apply thresholds: padj < 0.05 and abs(log2FoldChange) >= 1

filtered_results <- results[!is.na(results$padj < 0.05) & abs(results$log2FC) >= 1, ]
filtered_results <- as.data.frame(filtered_results) 
write.csv(filtered_results, '/home1/aditis54/490_cluster/analysis_data/SKCM_DESeq2_metastasis_filtered_results.csv')

# View filtered results
head(filtered_results)
```

```{r}
filtered_results$"-log10(padj)" <- log10(filtered_results$padj) 
head(filtered_results)

EnhancedVolcano(filtered_results,
                lab = filtered_results$gene_name,
                x = 'log2FC',
                y = 'X.log10.padj.',
                title = 'Sample Definition: Metastatic vs Non-metastatic',
                pointSize = 1.0,
                labSize = 5.0, 
                xlim = c(-5, 5), 
                ylim = c(0, 1))
                

```
3. Naive differential methylation analysis (volcano plot)
```{r}
if (!require("sesameData", quietly = TRUE))
BiocManager::install("sesameData")

if (!require("sesame", quietly = TRUE))
BiocManager::install("sesame")

if (!require("limma", quietly = TRUE))
BiocManager::install("limma")

install.packages("ggplot2")
```
```{r}
library(TCGAbiolinks)
library(sesame)
library(sesameData)
library(limma)
library(ggplot2)
```

```{r setup}
knitr::opts_knit$set(root.dir = normalizePath("/home1/aditis54/490_cluster/analysis_data"))
```

```{r}
query <- GDCquery(project = "TCGA-SKCM",
                  data.category = "DNA Methylation",
                  data.type = "Methylation Beta Value",
                  platform = "Illumina Human Methylation 450",
                  data.format = "Dataframe")
#GDCdownload(query)
methylation450 <- GDCprepare(query)
```
```{r}
methylation_clinical <- as.data.frame(methylation450@colData)
column_mask <- ifelse(colnames(methylation_clinical) %in% c('treatments', 'primary_site', 'disease_type'), F, T)
methylation_clinical <- methylation_clinical[,column_mask]
# write.csv(methylation_clinical, "/home1/aditis54/490_cluster/analysis_data/skcm_methylation_clinical.csv")

betas <- as.data.frame(methylation450@assays@data@listData)
cpg_sites <- as.data.frame(methylation450@rowRanges@elementMetadata)
site_mask <- !grepl('-', cpg_sites$gene) & !grepl(';', cpg_sites$gene) & !is.na(cpg_sites$gene) & complete.cases(betas)
betas <- betas[site_mask,]
cpg_sites <- cpg_sites[site_mask,]

#write.csv(betas, "/home1/aditis54/490_cluster/analysis_data/skcm_methylation_betas.csv")
#write.csv(cpg_sites, "/home1/aditis54/490_cluster/analysis_data/skcm_methylation_cpg_sites.csv")

```

```{r}
methylation_clinical <- read.csv('/home1/aditis54/490_cluster/analysis_data/skcm_methylation_clinical.csv')
betas <- read.csv('/project/rohs_1070/analysis_data/skcm_methylation_betas.csv')
cpg_sites <- read.csv('/project/rohs_1070/analysis_data/skcm_cpg_sites.csv')
```

```{r}

methylation_clinical$tumor <- ifelse(methylation_clinical$definition == "Metastatic" | methylation_clinical$definition == "Additional Metastatic", TRUE, 
                                     ifelse(methylation_clinical$definition == "Primary solid Tumor" | methylation_clinical$definition == "Solid Tissue Normal", FALSE, NA))

betas$X <- NULL

colnames(betas) <- methylation_clinical$barcode
betas <- betas[, methylation_clinical$barcode, drop = FALSE]

mval <- t(apply(betas, 1, function(x) log2(x/(1-x))))

design <- model.matrix(~ tumor, data = methylation_clinical)
fit <- lmFit(mval, design)
fit2 <- eBayes(fit)

```
```{r}
dat <- data.frame(foldchange = fit[["coefficients"]][,2], logPvalue =  -log10(p.adjust(fit2[["p.value"]][,2],method='BY')), geneName = cpg_sites$gene)
dat$threshold <- as.factor(abs(dat$foldchange) < 1)

#Visualization
cols <- c("TRUE" = "grey", "FALSE" = "blue")
ggplot(data=dat, aes(x=foldchange, y = logPvalue, color=threshold)) +
  geom_point(alpha=.2, size=0.6) +
  scale_colour_manual(values = cols) +
  geom_vline(xintercept = 1, colour="#990000", linetype="dashed") + 
  geom_vline(xintercept = - 1, colour="#990000", linetype="dashed") +
  geom_hline(yintercept = 2, colour = "#990000", linetype="dashed") +
  theme(legend.position="none") +
  xlab("Fold Change") +
  ylab("-log10 p value") +
  theme_bw() +
  theme(legend.position = "none")
```
All the points on the top right represent cpg sites that are hyper-methylated in metastatic patients. The ones on the left represent cpg sites that are under-methylated in metastatic patients relative to patients with primary tumors or normal tissue. Points near the center or below the significance threshold line are not statistically or biologically significant.

Part 4: Direct comparison of methylation status to transcriptional activity across non-metastatic vs metastatic patients

```{r}
# insert the deseq results here
results[, "-log10(padj)"] <- -log10(results$padj)
rna_genes_masked <- rna_genes[rna_genes$gene_id %in% results$gene_id, ]
rna_genes_masked <- rna_genes_masked[match(results$gene_id, rna_genes_masked$gene_id), ]
results[, "gene_name"] <- rna_genes_masked$gene_name
```
```{r}
upregulated <- results[(results$log2FC > 3), 'gene_name']
hypermethylated <- dat[dat$foldchange < 2, 'geneName']
interest_genes <- intersect(upregulated, hypermethylated)
interest_genes
```

```{r}
GENE<-"FCER2"
# change, must do for 10 intersect_genes
# CDH10, TRHDE, SLC17A6, FCER2, CLEC4M, PRG4, CR2, NRK, EDN3, MAEL

gene_counts_mask <- rna_genes$gene_name == GENE
gene_betas_mask <- cpg_sites$gene == GENE

rna_clinical_tumor <- rna_clinical$definition == "Metastatic"
methylation_clinical_tumor <- methylation_clinical$definition == "Metastatic"

rna_clinical_normal <- rna_clinical$definition == "Primary solid Tumor"
methylation_clinical_normal <- methylation_clinical$definition == "Primary solid Tumor"

rna_tumor <- as.numeric(rna_counts[gene_counts_mask, rna_clinical_tumor])
methylation_tumor <- (betas[gene_betas_mask, methylation_clinical_tumor])

rna_normal <- as.numeric(rna_counts[gene_counts_mask, rna_clinical_normal])
methylation_normal <- (betas[gene_betas_mask, methylation_clinical_normal])
```
```{r}
boxplot(rna_normal, rna_tumor, main = GENE, xlab='Group', ylab='Counts', names=c('Primary solid Tumor', 'Metastatic'))
```
```{r}
plot(1:dim(methylation_normal)[1], apply(methylation_normal, 1, mean), col='blue', type='h', lwd=3, main = GENE, ylab='Average Beta', xlab='CpG sites')
lines(1.3:(dim(methylation_normal)[1]+0.3), apply(methylation_tumor, 1, mean), col='red', type='h', lwd=3)
```



